web/src/features/datasets/server/service.ts

```diff
@@ -0,0 +1,329 @@
+          run_id String,  
+          run_item_id String,
+          dataset_id String,      
+          trace_id String,
+          observation_id Nullable(String)
+      )  ENGINE = Memory
+
+  `;
+  await commandClickhouse({
+    query,
+    params: { tableName },
+    clickhouseConfigs: { session_id: clickhouseSession },
+  });
+};
+
+export const deleteTempTableInClickhouse = async (tableName: string) => {
+  const query = `
+      DROP TABLE IF EXISTS ${tableName}
+  `;
+  await commandClickhouse({ query, params: { tableName } });
+};
+
+export const getDatasetRunsFromPostgres = async (
+  input: DatasetRunsTableInput,
+) => {
+  return await prisma.$queryRaw<PostgresDatasetRun[]>(
+    Prisma.sql`
+      SELECT
+        runs.id as run_id,
+        runs.name as run_name,
+        runs.description as run_description,
+        runs.metadata as run_metadata,
+        runs.created_at as run_created_at,
+        runs.updated_at as run_updated_at,
+        JSON_AGG(JSON_BUILD_OBJECT(
+          'trace_id', ri.trace_id,
+          'observation_id', ri.observation_id,
+          'ri_id', ri.id
+        )) AS run_items
+      FROM
+        datasets d
+        JOIN dataset_runs runs ON d.id = runs.dataset_id AND d.project_id = runs.project_id
+        JOIN dataset_run_items ri ON ri.dataset_run_id = runs.id
+          AND ri.project_id = runs.project_id
+      WHERE
+        d.id = ${input.datasetId}
+        AND d.project_id = ${input.projectId}
+      GROUP BY runs.id, runs.name, runs.description, runs.metadata, runs.created_at, runs.updated_at
+      LIMIT ${input.limit}
+      OFFSET ${input.page * input.limit}
+    `,
+  );
+};
+
+const getScoresFromTempTable = async (
+  input: DatasetRunsTableInput,
+  tableName: string,
+  clickhouseSession: string,
+) => {
+  const query = `
+      SELECT 
+        s.*,
+        tmp.run_id
+      FROM ${tableName} tmp JOIN scores s 
+        ON tmp.project_id = s.project_id 
+        AND tmp.observation_id = s.observation_id 
+        AND tmp.trace_id = s.trace_id
+      WHERE s.project_id = {projectId: String}
+      AND tmp.project_id = {projectId: String}
+      AND tmp.dataset_id = {datasetId: String}
+      ORDER BY s.event_ts DESC
+      LIMIT 1 BY s.id, s.project_id
+  `;
+
+  const rows = await queryClickhouse<
+    FetchScoresReturnType & { run_id: string }
+  >({
+    query: query,
+    params: {
+      projectId: input.projectId,
+      datasetId: input.datasetId,
+    },
+    clickhouseConfigs: { session_id: clickhouseSession },
+  });
+
+  return rows.map((row) => ({ ...convertToScore(row), run_id: row.run_id }));
+};
+
+const getObservationLatencyAndCostForDataset = async (
+  input: DatasetRunsTableInput,
+  tableName: string,
+  clickhouseSession: string,
+) => {
+  const query = `
+      WITH agg AS (
+      SELECT
+        run_id,
+        dateDiff('milliseconds', start_time, end_time) AS latency_ms,
+        total_cost AS cost
+      FROM ${tableName} tmp JOIN observations o 
```

Comment:
Does it make a difference whether we join the tmp table from right? That way we may avoid an in-memory hashmap of the observations table.

Reply:
thx